File: forge-core/internal/platformcorecontract/event.go
package platformcorecontract

// ValidateEventEnvelope validates one supplied event without appending it.
func ValidateEventEnvelope(value *EventEnvelope) error {
	_, err := CanonicalEventEnvelopeJSON(value)
	return err
}

// CanonicalEventEnvelopeJSON returns exact compact canonical v1 bytes.
func CanonicalEventEnvelopeJSON(value *EventEnvelope) ([]byte, error) {
	if err := validateEventFields(value); err != nil {
		return nil, withRejection(err, rejectionValueInvalid)
	}
	canonical, err := typedCanonical(value, maxEnvelopeBytes)
	return canonical, withRejection(err, rejectionValueInvalid)
}

// DecodeCanonicalEventEnvelope accepts only one exact canonical v1 event.
func DecodeCanonicalEventEnvelope(data []byte) (*EventEnvelope, error) {
	var value EventEnvelope
	if err := decodeTypedCanonical(data, maxEnvelopeBytes, &value); err != nil {
		return nil, withRejection(err, rejectionDocumentInvalid)
	}
	if err := validateExactTypedDocument(data, &value, maxEnvelopeBytes, "EventEnvelope"); err != nil {
		return nil, err
	}
	if err := validateEventFields(&value); err != nil {
		return nil, withRejection(err, rejectionValueInvalid)
	}
	return &value, nil
}

// EventEnvelopeSHA256 returns a domain-separated conformance digest.
func EventEnvelopeSHA256(value *EventEnvelope) (string, error) {
	canonical, err := CanonicalEventEnvelopeJSON(value)
	if err != nil {
		return "", err
	}
	return observationDigest(eventDigestDomain, canonical), nil
}

func validateEventFields(value *EventEnvelope) error {
	if value == nil {
		return reject(rejectionDocumentInvalid, "EventEnvelope is required")
	}
	view := eventEnvelopeValidationView(value)
	if err := validateEnvelopeValues(view); err != nil {
		return err
	}
	if err := validateEventValues(value); err != nil {
		return err
	}
	if err := validateEnvelopeReferences(view); err != nil {
		return err
	}
	if err := validateEventReferences(value); err != nil {
		return err
	}
	if err := validateEnvelopeRelations(view); err != nil {
		return err
	}
	return validateEventRelations(value)
}

func validateEventValues(value *EventEnvelope) error {
	if err := validateTypedID(value.EventID, "evt", "event_id"); err != nil {
		return err
	}
	if err := validateEntityRef(value.AggregateRef, "aggregate_ref"); err != nil {
		return err
	}
	if value.AggregateVersion < 1 || value.Sequence < 1 {
		return reject(rejectionValueInvalid, "aggregate_version and sequence must be positive")
	}
	if err := validateUnixMS(value.OccurredAtUnixMS, "occurred_at_unix_ms"); err != nil {
		return err
	}
	if _, err := parseSourceComponent(string(value.SourceComponent)); err != nil {
		return err
	}
	if value.SourceSnapshotRef != nil {
		return validateEntityRef(*value.SourceSnapshotRef, "source_snapshot_ref")
	}
	return nil
}

func validateEventReferences(value *EventEnvelope) error {
	if !scopeContains(value.ScopeRef, value.AggregateRef) {
		return reject(rejectionReferenceMismatch, "aggregate_ref is not represented by scope_ref")
	}
	scoped := value.ScopeRef.ProjectSnapshotID
	if scoped == nil && value.SourceSnapshotRef == nil {
		return nil
	}
	if scoped == nil || value.SourceSnapshotRef == nil ||
		value.SourceSnapshotRef.EntityType != entityProjectSnapshot ||
		value.SourceSnapshotRef.EntityID != *scoped {
		return reject(rejectionReferenceMismatch, "source_snapshot_ref must exactly match scoped project_snapshot_id")
	}
	return nil
}

func validateEventRelations(value *EventEnvelope) error {
	if !sameMessageSuffix(value.MessageID, value.EventID) {
		return reject(rejectionRelationMismatch, "message_id and event_id suffixes must match")
	}
	if value.PayloadArtifactRef != nil &&
		value.PayloadArtifactRef.CreatedAtUnixMS > value.OccurredAtUnixMS {
		return reject(rejectionRelationMismatch, "event payload artifact cannot be created after occurrence time")
	}
	return nil
}

func eventEnvelopeValidationView(value *EventEnvelope) envelopeValidationView {
	return envelopeValidationView{
		canonicalization: value.Canonicalization, causationID: value.CausationID,
		correlationID: value.CorrelationID, envelopeVersion: value.EnvelopeVersion,
		extensions: value.Extensions, messageID: value.MessageID, payload: value.Payload,
		payloadArtifact: value.PayloadArtifactRef, schemaName: value.SchemaName,
		schemaVersion: value.SchemaVersion, scope: value.ScopeRef, actor: value.ActorRef,
	}
}

File: forge-runtime/crates/domain/src/platform_core_contract/event.rs
use super::{
    EntityType, EventEnvelope, MAX_ENVELOPE_BYTES, MAX_UNIX_MILLISECONDS,
    PlatformCoreContractError, RejectionCode, SourceComponent, envelope, identity, invalid,
    references, reject, wire,
};

/// Validates a supplied event without appending or applying it.
///
/// # Errors
///
/// Returns an error for malformed identity, scope, payload, sequence, or bounds.
pub fn validate_event_envelope(value: &EventEnvelope) -> Result<(), PlatformCoreContractError> {
    let view = envelope_view(value);
    envelope::validate_common_values(&view)?;
    validate_event_values(value)?;
    envelope::validate_common_references(&view)?;
    validate_event_references(value)?;
    envelope::validate_common_relations(&view)?;
    validate_event_relations(value)?;
    wire::canonical_typed(value, MAX_ENVELOPE_BYTES).map(|_| ())
}

fn validate_event_values(value: &EventEnvelope) -> Result<(), PlatformCoreContractError> {
    identity::validate_typed_id(&value.event_id, "evt", "event_id")?;
    references::validate_entity_ref(&value.aggregate_ref, "aggregate_ref")?;
    if value.aggregate_version < 1 || value.sequence < 1 {
        return Err(invalid("aggregate_version and sequence must be positive"));
    }
    if !(1..=MAX_UNIX_MILLISECONDS).contains(&value.occurred_at_unix_ms) {
        return Err(invalid(format!(
            "occurred_at_unix_ms must be in 1..{MAX_UNIX_MILLISECONDS}"
        )));
    }
    if matches!(&value.source_component, SourceComponent::Unknown(_)) {
        return Err(reject(
            RejectionCode::ValueInvalid,
            "source_component is unsupported",
        ));
    }
    if let Some(source) = &value.source_snapshot_ref {
        references::validate_entity_ref(source, "source_snapshot_ref")?;
    }
    Ok(())
}

fn validate_event_references(value: &EventEnvelope) -> Result<(), PlatformCoreContractError> {
    if !references::scope_contains(&value.scope_ref, &value.aggregate_ref) {
        return Err(reject(
            RejectionCode::ReferenceMismatch,
            "aggregate_ref is not represented by scope_ref",
        ));
    }
    validate_event_snapshot(value)
}

fn validate_event_snapshot(value: &EventEnvelope) -> Result<(), PlatformCoreContractError> {
    match (
        value.scope_ref.project_snapshot_id.as_deref(),
        value.source_snapshot_ref.as_ref(),
    ) {
        (None, None) => Ok(()),
        (Some(scoped), Some(source))
            if source.entity_type == EntityType::ProjectSnapshot && source.entity_id == scoped =>
        {
            Ok(())
        }
        _ => Err(reject(
            RejectionCode::ReferenceMismatch,
            "source_snapshot_ref must exactly match scoped project_snapshot_id",
        )),
    }
}

fn validate_event_relations(value: &EventEnvelope) -> Result<(), PlatformCoreContractError> {
    if !identity::same_message_suffix(&value.message_id, &value.event_id) {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "message_id and event_id suffixes must match",
        ));
    }
    if value
        .payload_artifact_ref
        .as_ref()
        .is_some_and(|artifact| artifact.created_at_unix_ms > value.occurred_at_unix_ms)
    {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "event payload artifact cannot be created after occurrence time",
        ));
    }
    Ok(())
}

fn envelope_view(value: &EventEnvelope) -> envelope::EnvelopeView<'_> {
    envelope::EnvelopeView {
        canonicalization: &value.canonicalization,
        causation_id: &value.causation_id,
        correlation_id: &value.correlation_id,
        envelope_version: value.envelope_version,
        extensions: &value.extensions,
        message_id: &value.message_id,
        payload: &value.payload,
        payload_artifact_ref: &value.payload_artifact_ref,
        schema_name: &value.schema_name,
        schema_version: value.schema_version,
        scope_ref: &value.scope_ref,
        actor_ref: &value.actor_ref,
    }
}
